File: packages/sdk-ts/src/robotics/capability.ts
/**
 * Physical capability scope for robots (Phase 5.3), TypeScript.
 *
 * Mirrors `vouch/robotics/capability.py`. Extends capability attenuation to the
 * physical world: max force and speed, a slower speed cap near humans, allowed
 * zones, and shift windows, carried in a signed credential so the bound is
 * cryptographically enforceable. A controller checks a proposed action against
 * the granted scope before actuating, and a delegated scope must attenuate
 * (narrow, never broaden) its parent.
 */

import type { Signer } from '../signer';
import { VC_CONTEXT_V2, VOUCH_CONTEXT_V1 } from '../vc';

export const PHYSICAL_SCOPE_TYPE = 'PhysicalCapabilityScope';

export interface PhysicalAction {
  forceN?: number;
  speedMps?: number;
  nearHumans?: boolean;
  zone?: string;
  timeHm?: string; // "HH:MM" local
}

export interface CheckResult {
  ok: boolean;
  reasons: string[];
}

export interface ShiftWindow {
  start: string;
  end: string;
}

function iso(d: Date): string {
  return d.toISOString().replace(/\.\d{3}Z$/, 'Z');
}

export interface BuildPhysicalScopeOptions {
  subjectDid: string;
  maxForceN?: number;
  maxSpeedMps?: number;
  maxSpeedNearHumansMps?: number;
  allowedZones?: string[];
  shiftWindows?: ShiftWindow[];
  validSeconds?: number;
  validFrom?: Date;
}

/** Build a signed PhysicalCapabilityScope credential. */
export async function buildPhysicalScopeCredential(
  signer: Signer,
  opts: BuildPhysicalScopeOptions
): Promise<Record<string, unknown>> {
  const scope: Record<string, unknown> = {};
  if (opts.maxForceN !== undefined) scope.maxForceN = opts.maxForceN;
  if (opts.maxSpeedMps !== undefined) scope.maxSpeedMps = opts.maxSpeedMps;
  if (opts.maxSpeedNearHumansMps !== undefined) {
    scope.maxSpeedNearHumansMps = opts.maxSpeedNearHumansMps;
  }
  if (opts.allowedZones !== undefined) scope.allowedZones = [...opts.allowedZones];
  if (opts.shiftWindows !== undefined) scope.shiftWindows = opts.shiftWindows.map((w) => ({ ...w }));

  const issued = opts.validFrom ?? new Date();
  const credential: Record<string, unknown> = {
    '@context': [VC_CONTEXT_V2, VOUCH_CONTEXT_V1],
    type: ['VerifiableCredential', PHYSICAL_SCOPE_TYPE],
    issuer: signer.getDid(),
    validFrom: iso(issued),
    credentialSubject: { id: opts.subjectDid, physicalScope: scope },
  };
  if (opts.validSeconds !== undefined) {
    credential.validUntil = iso(new Date(issued.getTime() + opts.validSeconds * 1000));
  }
  return signer.attachProof(credential);
}

function inWindow(hm: string, w: { start?: string; end?: string }): boolean {
  return (w.start ?? '00:00') <= hm && hm <= (w.end ?? '23:59');
}

/** Check a proposed physical action against a physical scope object. */
export function checkPhysicalAction(
  scope: Record<string, any>,
  action: PhysicalAction
): CheckResult {
  const reasons: string[] = [];

  if (action.forceN !== undefined && 'maxForceN' in scope) {
    if (action.forceN > scope.maxForceN) {
      reasons.push(`force_exceeded: ${action.forceN}N > ${scope.maxForceN}N`);
    }
  }

  if (action.speedMps !== undefined) {
    let cap = scope.maxSpeedMps;
    if (action.nearHumans && 'maxSpeedNearHumansMps' in scope) {
      cap = scope.maxSpeedNearHumansMps;
    }
    if (cap !== undefined && action.speedMps > cap) {
      const label = action.nearHumans ? 'near_humans ' : '';
      reasons.push(`${label}speed_exceeded: ${action.speedMps} m/s > ${cap} m/s`);
    }
  }

  if (action.zone !== undefined && 'allowedZones' in scope) {
    if (!scope.allowedZones.includes(action.zone)) {
      reasons.push(`zone_not_allowed: ${action.zone}`);
    }
  }

  if (action.timeHm !== undefined && 'shiftWindows' in scope) {
    const windows = scope.shiftWindows as ShiftWindow[];
    if (windows.length && !windows.some((w) => inWindow(action.timeHm as string, w))) {
      reasons.push(`outside_shift_window: ${action.timeHm}`);
    }
  }

  return { ok: reasons.length === 0, reasons };
}

/**
 * True if `child` is a valid attenuation of `parent`: never broader on any
 * physical dimension. Numeric caps may only shrink; allowed zones may only be a
 * subset; shift windows must each fit inside some parent window.
 */
export function attenuates(
  parent: Record<string, any>,
  child: Record<string, any>
): boolean {
  for (const key of ['maxForceN', 'maxSpeedMps', 'maxSpeedNearHumansMps']) {
    if (key in parent) {
      if (!(key in child)) return false;
      if (child[key] > parent[key]) return false;
    }
  }

  if ('allowedZones' in parent) {
    const pZones = new Set<string>(parent.allowedZones);
    const cZones: string[] = child.allowedZones ?? [];
    if (cZones.length === 0 || !cZones.every((z) => pZones.has(z))) return false;
  }

  if ('shiftWindows' in parent) {
    const pWindows = parent.shiftWindows as ShiftWindow[];
    for (const cw of (child.shiftWindows ?? []) as ShiftWindow[]) {
      const fits = pWindows.some(
        (pw) => (pw.start ?? '00:00') <= (cw.start ?? '00:00') && (cw.end ?? '23:59') <= (pw.end ?? '23:59')
      );
      if (!fits) return false;
    }
  }

  return true;
}

File: go-sidecar/robotics/capability.go
// Physical capability scope for robots (Phase 5.3), Go.
//
// Mirrors vouch/robotics/capability.py and the TypeScript SDK. Extends capability
// attenuation to the physical world: max force and speed, a slower speed cap near
// humans, allowed zones, and shift windows, carried in a signed credential so the
// bound is cryptographically enforceable. A controller checks a proposed action
// against the granted scope before actuating, and a delegated scope must attenuate
// (narrow, never broaden) its parent.
//
// CheckPhysicalAction and Attenuates accept a scope as map[string]any, which is
// both how the builder assembles it and how it arrives after a cross-language
// JSON verify (numbers as float64, arrays as []any of map[string]any). The
// coercion helpers below handle both shapes so the same scope verified in Python
// or TypeScript checks identically in Go.
package robotics

import (
	"fmt"
	"time"

	"github.com/vouch-protocol/vouch/go-sidecar/signer"
)

// PhysicalScopeType is the credential type for a physical capability scope.
const PhysicalScopeType = "PhysicalCapabilityScope"

// PhysicalAction is a proposed actuation to check against a scope. Pointer
// numerics distinguish "not specified" from a meaningful zero (0 N is a real
// force bound).
type PhysicalAction struct {
	ForceN     *float64
	SpeedMps   *float64
	NearHumans bool
	Zone       string // "" = not specified
	TimeHm     string // "HH:MM" local; "" = not specified
}

// CheckResult is the outcome of CheckPhysicalAction.
type CheckResult struct {
	OK      bool
	Reasons []string
}

// ShiftWindow is an allowed time-of-day window, "HH:MM" start and end.
type ShiftWindow struct {
	Start string
	End   string
}

// BuildPhysicalScopeOptions configures BuildPhysicalScopeCredential. Pointer
// numerics omit the field when nil.
type BuildPhysicalScopeOptions struct {
	SubjectDID            string
	MaxForceN             *float64
	MaxSpeedMps           *float64
	MaxSpeedNearHumansMps *float64
	AllowedZones          []string      // nil omits the field
	ShiftWindows          []ShiftWindow // nil omits the field
	ValidSeconds          int           // 0 omits validUntil
	ValidFrom             time.Time     // zero uses now
}

// BuildPhysicalScopeCredential builds a signed PhysicalCapabilityScope credential.
func BuildPhysicalScopeCredential(s *signer.Signer, opts BuildPhysicalScopeOptions) (map[string]any, error) {
	scope := map[string]any{}
	if opts.MaxForceN != nil {
		scope["maxForceN"] = *opts.MaxForceN
	}
	if opts.MaxSpeedMps != nil {
		scope["maxSpeedMps"] = *opts.MaxSpeedMps
	}
	if opts.MaxSpeedNearHumansMps != nil {
		scope["maxSpeedNearHumansMps"] = *opts.MaxSpeedNearHumansMps
	}
	if opts.AllowedZones != nil {
		zones := make([]any, len(opts.AllowedZones))
		for i, z := range opts.AllowedZones {
			zones[i] = z
		}
		scope["allowedZones"] = zones
	}
	if opts.ShiftWindows != nil {
		ws := make([]any, len(opts.ShiftWindows))
		for i, w := range opts.ShiftWindows {
			ws[i] = map[string]any{"start": w.Start, "end": w.End}
		}
		scope["shiftWindows"] = ws
	}

	issued := opts.ValidFrom
	if issued.IsZero() {
		issued = time.Now().UTC()
	}
	cred := map[string]any{
		"@context":          []any{vcContextV2, vouchContextV1},
		"type":              []any{"VerifiableCredential", PhysicalScopeType},
		"issuer":            s.DID(),
		"validFrom":         iso(issued),
		"credentialSubject": map[string]any{"id": opts.SubjectDID, "physicalScope": scope},
	}
	if opts.ValidSeconds > 0 {
		cred["validUntil"] = iso(issued.Add(time.Duration(opts.ValidSeconds) * time.Second))
	}
	return s.AttachProof(cred)
}

// CheckPhysicalAction checks a proposed physical action against a physical scope
// object, returning ok plus a list of reasons for any violation.
func CheckPhysicalAction(scope map[string]any, action PhysicalAction) CheckResult {
	var reasons []string

	if action.ForceN != nil {
		if cap, ok := scopeNum(scope, "maxForceN"); ok && *action.ForceN > cap {
			reasons = append(reasons, fmt.Sprintf("force_exceeded: %vN > %vN", *action.ForceN, cap))
		}
	}

	if action.SpeedMps != nil {
		cap, has := scopeNum(scope, "maxSpeedMps")
		if action.NearHumans {
			if nh, ok := scopeNum(scope, "maxSpeedNearHumansMps"); ok {
				cap, has = nh, true
			}
		}
		if has && *action.SpeedMps > cap {
			label := ""
			if action.NearHumans {
				label = "near_humans "
			}
			reasons = append(reasons, fmt.Sprintf("%sspeed_exceeded: %v m/s > %v m/s", label, *action.SpeedMps, cap))
		}
	}

	if action.Zone != "" {
		if zonesAny, ok := scope["allowedZones"]; ok {
			if !containsStr(toStrSlice(zonesAny), action.Zone) {
				reasons = append(reasons, "zone_not_allowed: "+action.Zone)
			}
		}
	}

	if action.TimeHm != "" {
		if wsAny, ok := scope["shiftWindows"]; ok {
			windows := toWindows(wsAny)
			if len(windows) > 0 {
				inAny := false
				for _, w := range windows {
					if inWindow(action.TimeHm, w) {
						inAny = true
						break
					}
				}
				if !inAny {
					reasons = append(reasons, "outside_shift_window: "+action.TimeHm)
				}
			}
		}
	}

	return CheckResult{OK: len(reasons) == 0, Reasons: reasons}
}

// Attenuates reports whether child is a valid attenuation of parent: never broader
// on any physical dimension. Numeric caps may only shrink; allowed zones may only
// be a subset; shift windows must each fit inside some parent window.
func Attenuates(parent, child map[string]any) bool {
	for _, key := range []string{"maxForceN", "maxSpeedMps", "maxSpeedNearHumansMps"} {
		if pv, ok := scopeNum(parent, key); ok {
			cv, has := scopeNum(child, key)
			if !has || cv > pv {
				return false
			}
		}
	}

	if pz, ok := parent["allowedZones"]; ok {
		pset := toStrSlice(pz)
		cz := toStrSlice(child["allowedZones"])
		if len(cz) == 0 {
			return false
		}
		for _, z := range cz {
			if !containsStr(pset, z) {
				return false
			}
		}
	}

	if pw, ok := parent["shiftWindows"]; ok {
		pWindows := toWindows(pw)
		for _, cw := range toWindows(child["shiftWindows"]) {
			cs, ce := windowBounds(cw)
			fits := false
			for _, pwin := range pWindows {
				ps, pe := windowBounds(pwin)
				if ps <= cs && ce <= pe {
					fits = true
					break
				}
			}
			if !fits {
				return false
			}
		}
	}

	return true
}

func inWindow(hm string, w ShiftWindow) bool {
	start, end := windowBounds(w)
	return start <= hm && hm <= end
}

func windowBounds(w ShiftWindow) (string, string) {
	start, end := w.Start, w.End
	if start == "" {
		start = "00:00"
	}
	if end == "" {
		end = "23:59"
	}
	return start, end
}

// scopeNum reads a numeric scope field, coercing the float64 that JSON decoding
// produces as well as native Go numerics.
func scopeNum(scope map[string]any, key string) (float64, bool) {
	v, ok := scope[key]
	if !ok {
		return 0, false
	}
	switch n := v.(type) {
	case float64:
		return n, true
	case float32:
		return float64(n), true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	}
	return 0, false
}

func toStrSlice(v any) []string {
	switch s := v.(type) {
	case []string:
		return s
	case []any:
		out := make([]string, 0, len(s))
		for _, e := range s {
			if str, ok := e.(string); ok {
				out = append(out, str)
			}
		}
		return out
	}
	return nil
}

func toWindows(v any) []ShiftWindow {
	switch ws := v.(type) {
	case []ShiftWindow:
		return ws
	case []any:
		out := make([]ShiftWindow, 0, len(ws))
		for _, e := range ws {
			if m, ok := e.(map[string]any); ok {
				w := ShiftWindow{}
				w.Start, _ = m["start"].(string)
				w.End, _ = m["end"].(string)
				out = append(out, w)
			}
		}
		return out
	}
	return nil
}

func containsStr(s []string, v string) bool {
	for _, x := range s {
		if x == v {
			return true
		}
	}
	return false
}
